### PR TITLE
Convenient Pagination API, Browser Support, Missing API fields

### PR DESCRIPTION
**Pull request by pyrogenic**
Several additions to Discojs that I've accumulated over the years building [Elephant](https://pyrogenic.github.io/elephant/). It's a browser-based tool to manage a Discogs collection and store listings. ([source](https://github.com/pyrogenic/elephant))

- Add an `all` API to Discojs that retrieves all pages of a paginated endpoint given the first page:
  ```ts
  function updateCollection() {
    return client.listItemsInFolder(0, undefined, {
      perPage: 100,
    }).then(((r) => client.all("releases", r, addToCollection)), setError);
  }
  ```
- Add a model for the error format used by Discogs to allow better error reporting
- Add a few missing fields to various responses (`notes` to `FolderReleases`, `name` to `Artist`, `avatar_url` to `User`)
- Fill out the model for marketplace listings, which did not include any of the owner-only fields
- Fix nesting of the model for per-track information
- Add the option to provide a `ResultCache` which may be used to skip retrieving results multiple times. (My app stores `GET` responses using IndexedDB.)
- Tweak the package scripts to work on Windows (I develop on several platforms)

Happy to make changes to allow the PR to be accepted.

### CODE REVIEW

**Comment by aknorw** [comment]
@pyrogenic Sorry for missing this one, and thanks for all the improvements!
Would it be possible to split this PR into multiple ones so it'd be easier to review?
Happy to help!

File: models/api.ts [hunk 1]
```diff
@@ -21,6 +21,29 @@ import { ExportItemIO } from './inventory'
...
+        msg: t.string,
+        type: t.string,
+      }),
+    ),
+  }),
+])
+
+export type ErrorResponse = t.TypeOf<typeof ErrorResponseIO>
+
+export type Pagination = t.TypeOf<typeof PaginationIO>
```
Comment (aknorw) [inline models/api.ts:41]: We already have a `Pagination` type, named this one `PaginationResponse`.

File: models/marketplace.ts [hunk 2]
```diff
@@ -70,7 +70,7 @@ export const ListingIO = t.intersection([
   ResourceURLIO,
   t.type({
     id: t.Integer,
-    status: makeEnumIOType(ListingStatusesEnum),
+    status: makeEnumIOType(InventoryStatusesEnum),
```
Comment (aknorw) [inline models/marketplace.ts:73]: I chose not to introduce this change yet, as this is only the case when users are authenticated. IMHO we should have 2 different codecs/types depending on whether the user is authenticated.

File: src/allNext.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import { IPaginated } from '../models/api'
+import type { Discojs } from './discojs'
+
+export class AllNext {
```
Comment (aknorw) [inline src/allNext.ts:4]: I felt this implementation was a bit hard to follow, as you first have to call a method, then call `client.all`.
Instead of doing that, I introduced dedicated `all` methods that returns async iterators:
* `getAllMasterVersions`
* `getAllArtistReleases`
* `getAllLabelReleases`
* `getAllRecentExports`
* `getAllInventoryForUser`
* `getAllInventory`
* `listAllOrders`
* `listAllItemsByReleaseForUser`
* `listAllItemsByRelease`
* `listAllItemsInFolderForUser`
* `listAllItemsInFolder`
* `getAllSubmissionsForUser`
* `getAllSubmissions`
* `getAllContributionsForUser`
* `getAllContributions`
* `getAllListsForUser`
* `getAllLists`
* `getAllWantlistForUser`
* `getAllWantlist`

The example you have in the PR description would become:
```ts
async function updateCollection() {
  for await (const { releases } of client.listAllItemsInFolder()) {
    for (const release of releases) {
      await addToCollection(release)
    }
  }
}
```

File: src/utils/fetch.ts [hunk 4]
```diff
@@ -22,6 +27,13 @@ const RATE_LIMIT_HEADER = 'X-Discogs-Ratelimit'
 /** Header set by Discogs API to indicate the number of remaining requests you are able to make in the existing rate limit window. */
 const RATE_LIMIT_REMAINING_HEADER = 'X-Discogs-Ratelimit-Remaining'
 
+let buffer: undefined | typeof Buffer
+try {
+  buffer = Buffer
+} catch {
+  buffer = undefined
+}
```
Comment (aknorw) [inline src/utils/fetch.ts:35]: What is the purpose of this?

**Review by aknorw — COMMENTED** [review]
@pyrogenic I spent some time yesterday introducing some changes initiated in this PR:
* Add missing fields (see comments for ones I did not add)
* Improve error reporting
* Introduce result cache
* Add allowUnsafeHeaders option
* Introduce dedicated methods to iterate through pages (see comment in the review about the implemenation difference)

Let me know what you think about them!

**Comment by pyrogenic** [comment]
@aknorw thanks so much for taking the time to review, and sorry for the delayed response. I'll take a closer look outside work hours!

... and apologies for throwing a couple years of tweaks up in a single PR!

File: models/marketplace.ts (hunk 2 above)
Comment (pyrogenic) [reply models/marketplace.ts:73]: Ah, I didn't know that difference, I only saw it not match.  Is it the user being authenticated at all, or the authenticated user being the owner of the listing?

I'll check out the code paths I encounter this and see what might make sense.

File: src/utils/fetch.ts (hunk 4 above)
Comment (pyrogenic) [reply src/utils/fetch.ts:35]: I'm guessing there was a situation in some browser where _accessing_ `Buffer` to calculate content length through. I honestly don't remember the details. Getting Chrome to be happy with those headers and their capitalization was a PITA. I myself would reject this without an explanatory comment about the circumstances of the error and the expected alternate behavior...

File: src/allNext.ts (hunk 3 above)
Comment (pyrogenic) [reply src/allNext.ts:4]: That seems much more user-friendly, if less magical. 😉

I probably chose my method based on writing the least code and exercising some TypeScript-fu.
